src/datasets/utils/beam_utils.py

```diff
@@ -1,52 +1,195 @@
+            batch is a basic unit used for storing data in the row group buffer.
+            A higher record batch size implies low granularity on a row group buffer
+            size. For configuring a row group size based on the number of records,
+            set ``row_group_buffer_size`` to 1 and use ``record_batch_size`` to
+            adjust the value.
+        codec (`str`, *optional*): The codec to use for block-level compression. Any string supported
+            by the pyarrow specification is accepted.
+        file_name_suffix (`str`, defaults to `""`): Suffix for the files written.
+        num_shards (`int`, defaults to `0`): The number of files (shards) used for output. If not set, the
+            service will decide on the optimal number of shards.
+            Constraining the number of shards is likely to reduce
+            the performance of a pipeline.  Setting this value is not recommended
+            unless you require a specific number of output files.
+        shard_name_template (`str`, *optional*): A template string containing placeholders for
+            the shard number and shard count. When constructing a filename for a
+            particular shard number, the upper-case letters 'S' and 'N' are
+            replaced with the 0-padded shard number and shard count respectively.
+            This argument can be '' in which case it behaves as if num_shards was
+            set to 1 and only one file will be generated. The default pattern used
+            is '-SSSSS-of-NNNNN' if None is passed as the shard_name_template.
+        mime_type (`str`, defaults to `"application/x-arrow"`): The MIME type to use for the produced files, if the filesystem
+            supports specifying MIME types.
+
+    Returns:
+        A `WriteToArrow` transform usable for writing.
+    """
+
+    def __init__(
+        self,
+        file_path_prefix: str,
+        schema: pa.Schema,
+        row_group_buffer_size: int = 64 * 1024 * 1024,
+        record_batch_size: int = 1000,
+        codec: Optional[str] = None,
+        file_name_suffix: str = "",
+        num_shards: int = 0,
+        shard_name_template: Optional[str] = None,
+        mime_type: str = "application/x-arrow",
+    ):
+        super().__init__()
+        self._schema = schema
+        self._row_group_buffer_size = row_group_buffer_size
+        self._record_batch_size = record_batch_size
+
+        self._sink = _IpcStreamSink(
+            file_path_prefix, schema, codec, file_name_suffix, num_shards, shard_name_template, mime_type
+        )
+
+    def expand(self, pcoll):
+        return (
+            pcoll
+            | ParDo(_RowDictionariesToArrowTable(self._schema, self._row_group_buffer_size, self._record_batch_size))
+            | Write(self._sink)
+        )
+
+    def display_data(self):
+        return {"sink_dd": self._sink, "row_group_buffer_size": str(self._row_group_buffer_size)}
+
+
+class _IpcStreamSink(filebasedsink.FileBasedSink):
```

Comment:
FileBasedSink also supports max_shard_size, which can be useful in our case since _prepare_split also has this argument

Reply:
Is [this](https://github.com/apache/beam/blob/62b7247a28e1f9f4fa23cfbe202f939a5634caa8/sdks/python/apache_beam/io/filebasedsink.py#L73) the param you are referring to? Unfortunately, it's not exposed in `_ParquetSink`, so we would also have to subclass `WriteToParquet` and `_ParquetSink` to address this.

Reply:
indeed, maybe let's ignore for now.

I also opened https://github.com/apache/beam/pull/23808 to add it to parquet sink

Reply:
I resolved this by adding these parameters to our own implementations of the Write transforms :).